Enhance pbs_config & pbs_stat to support snapshots instead of diags

agrawalravi90 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
- pbs_config has the capability of taking in output from pbs_diag and configuring the local PBS system to "mimic" the diag captured (creates nodes, queues, scheduling policies etc. to make the local system appear as if it's the diag). Since we replaced pbs_diag with pbs_snapshot, it makes it necessary to enhance pbs_config to work with output of pbs_snapshot instead.

- pbs_config right now does not create users/groups corresponding to those which existed at the customer site. In order to debug esoteric issues, sometimes it's useful to submit exactly the kind of jobs which were being submitted at the site. So, it'd also be useful to have pbs_config --as-snap accept a sub-option --acct-logs to take an accounting trace from a site and create the users & groups.

#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->
- Removed --as-diag option
- Added new option --as-snap to pass snapshots to pbs_config for it to re-create the PBS configuration parsed from the snapshot
- Added new option --acct-logs to pass in accounting logs for pbs_config to create users/groups from the workload trace

#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->
https://pbspro.atlassian.net/wiki/spaces/PD/pages/1279590401/Enhance+pbs+config+to+support+pbs+snapshot


#### Attach Test Logs or Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
[testconfig.log](https://github.com/PBSPro/pbspro/files/3198662/testconfig.log)




<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -3512,12 +3513,12 @@ def __init__(self, name=None, attrs={}, defaults={}, pbsconf_file=None,
         # If diag is given then bypass parsing pbs.conf
         if self.has_diag:
```
I suggest renaming `diag` with `snap` in this file like `self.has_diag` with `self.has_snap`, `self.diag` with `self.snap` etc...

agrawalravi90 replied:
I was being lazy :) I'll rename them, thanks

hirenvadalia commented on test/tests/functional/pbs_config.py:
```diff
@@ -0,0 +1,119 @@
+from tests.functional import *
+
+
+class TestPBSConfig(TestFunctional):
```
shouldn't this file go to selftest?

agrawalravi90 replied:
No, pbs_config is a tool/command, just like qmgr, qstat etc., so it should be in Functional. If we write a test especially to test PTL's 'clusterize' interface, then that would be in selftest as it tests PTL itself.

agrawalravi90 commented:
@hirenvadalia I've addressed your comment. I ended up replacing diag with snapshot for pbs_stat command as well. Please let me know what you think. Thanks

bhroam commented on test/tests/functional/pbs_config.py:
```diff
@@ -0,0 +1,124 @@
+
+        # Cleanup
+        self.du.rm(path=snap_dir, recursive=True, force=True)
+        self.du.rm(path=output_tar, force=True)
```
Put this in tearDown().  If the above expect()s or asserts fail, the test will fail and this won't happen.

bhroam commented:
Travis failed to a pep8 error.  One line is too long.

bhroam commented on test/tests/functional/pbs_config.py:
```diff
@@ -81,7 +81,7 @@ def test_config_as_snap(self):
         outdir = pwd.getpwnam(self.du.get_current_user()).pw_dir
         snap_cmd = [pbs_snapshot_path, "-o " + outdir, "--with-sudo"]
-        ret = self.du.run_cmd(cmd=snap_cmd, logerr=False)
+        ret = self.du.run_cmd(cmd=snap_cmd, logerr=False, as_script=True)
```
Just curious, why do you need as_script=True?  You are just running a normal command.  You aren't doing anything that requires the shell the parse it.

agrawalravi90 replied:
it fails without it. I think it's because we run some commands with sudo when --with-sudo is provided.

agrawalravi90 commented on test/fw/bin/pbs_stat:
```diff
@@ -83,7 +83,7 @@ def usage():
     msg += ['-c: counts number of matching items\n']
     msg += ['-C: counts grand total of given attribute values in complex\n']
-    msg += ['-d <pbs_diag>: path to diag directory\n']
+    msg += ['-d <pbs_snapshot>: path to snap directory\n']
```
-s is taken, i could make it -S, does that sound better?

bhroam replied:
Either that or remove the short argument and just use the long argument.

agrawalravi90 commented on test/fw/bin/pbs_stat:
```diff
@@ -149,24 +149,24 @@ def usage():
+            snap = snap.replace(".", "")
+            snap = snap.replace("/", "_")
+            snap = snap.replace("snapshot_", "")
+            if snap[0] in ('/', '_'):
```
not my code :) I didn't modify anything besides renaming diag to snap. I'll remove '/'

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -1606,9 +1606,8 @@ def file_to_dictlist(self, file=None, attribs=None, id=None):
-            f = open(file, 'r')
-            lines = f.readlines()
-            f.close()
+            with open(file, 'r') as f:
```
file is a python builtin.  Consider using filename like you do above

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -1932,9 +1930,8 @@ def file_to_batch(self, file=None):
-            l = f.readlines()
-            f.close()
+            with open(file, 'r') as f:
+                l = f.readlines()
```
'l' could be a misleading variable name.  Consider using something different.

agrawalravi90 commented:
@bhroam  I've addressed your comments, please let me know if it looks good. Thanks!

bhroam commented:
Once you decide what you want to do with the argument, I'm fine with the code.  I'm leaning to just making it a long argument and foregoing the short.

agrawalravi90 commented:
@bhroam thanks, I've renamed the options of both commands to "--snap". I've edited the design doc to reflect the same. Please let me know if it looks okay now. @hirenvadalia requesting you to review the latest commit once as well. Thanks.

bhroam commented:
Once @hirenvadalia takes another look at the changes, if you could squash and rebase once (I'll take it from there).  It's been a few days since the first commit.  Something might have changed.

agrawalravi90 commented:
Thanks @bhroam . @hirenvadalia please review the last commit whenever you get some time. Thanks

agrawalravi90 commented:
Thanks @hirenvadalia !

@bhroam I've done a rebase, uploaded new test logs, please merge this in if all looks good, thanks

bhroam commented:
Thanks @agrawalravi90.  I merged in the PR.
